### ISSUE TITLE
Auto-renaming FX Mixer strips (and assigning newly added instruments to unused mixer channels)

### ISSUE THREAD

**Issue by unfa** [LMMS/lmms#48]
A lot of time is wasted on manually duplicating names from instruments to mixer strips these instruments are assigned to. Also assigning the instruments to channels takes some time and could be automated as well. This will sort the problem out.

I think we could make this optional in the preferences menu and allow users to customize how it works a bit

```
 [x] Automatically assign an unused FX Mixer channel to every new instrument in the project
 [x] Automatically name mixer strips by assigned instruments
 [x] Allow shortening names (Hihat 1 → Ht1, Bass → Bs, Acid Lead with Delay → AdLawDy etc.)
 [200 ms] Delay time before performing auto-rename to save CPU power (set 0 for disabled)
 [   +   ] String used to join multiple instrument names (joining "Pad 1" and "Pad 2" gives "Pad 1 + Pad 2")
```

As for shortening names, Ardour has this feature for Plugins Stack and Mixer strips (while in small mode) - maybe we could use their code for this, it works well).

I think the renaming should be done by the mixer, but triggered by instruments.

Mixer should have a method to "update auto-rename" that is run for a single mixer (int I guess).
For example if an instrument named "Bass" was switched using mouse wheel from channel 0 (master) to 5, every time the change occurs (for values 0, 1, 2, 3, 4, 5) the instrument should report change for two mixers strips:

```
 mixer.updateAutoRename(previously_used_mixer_strip_number);
 mixer.updateAutoRename(now_used_mixer_strip_number);
```

If this is too heavy on the CPU we could add a timer to launch this only after a while to avoid name-sweeping through the mixer (although it would look nice) - let's users set it up too.

The mixer should have a way to know, which instruments are connected to any given FX channel. Now it would look at that table:

```
 int mixer.updateAutoRename(int index)
 {
      if (mixer.channels[index].assignedInstruments.count > 1)
      {
          /* shorten the names and join them with a string that user sets up in preferences namely "mixerAutoRenameJointString" */
      }
      else
      {
           mixer.channels[index].name = mixer.channels[index].assignedInstruments[0].name;
      }
 }
```

Every strip should also have a "pre-auto-rename" string var to store whatever the autorename replaced. Also two bool variables for "use-auto-rename" and "mark-auto-renamed" should be present as if the user manually changes the name, the auto-rename should not occur anymore, unless he manually turn it on (Rightclick on FX-mixer strip > "[ ] Automatically rename this mixer strip according to assigned instruments". However automatically chosen names should be marked with a slightly different colour or font (maybe auto-renames should be in Italics?) - so users can see in a blink of an eye witch names are automatic, an which are not.

Another thing is to request mixer to assign an unused mixer strip to every newly added instrument. Then if the user renames it, the auto-rename feature will do the rest.

If mixer strips will be given a color variable, there can be another switch to "Change colors too" when user changes a color of a instrument (if it's added) then the mixer strip's color would change accordingly - if several instruments are assigned to one channel, colors should be blended (for example RED and YELLOW will give us ORANGE).

**Comment by tresf** [comment]
> This idea is continued here: #1215

Then that's where it can remain.  Closing as duplicate of #1215
